### ISSUE TITLE
If you reuse a class based on AbstractJavadocCheck, you can get stale error messages with wrong information

### ISSUE THREAD

**Issue by Vampire** [checkstyle/checkstyle#2326]
We use SonarQube with CheckStyle checks.
We recently activated the AtclauseOrderCheck.
But this caused the whole SQ analysis to fail.

The reason is the following:

We have a file A.java with JavaDoc:

```
   /**
    * get the sorted list of tariffs to display in the combo box
    *
    * @param listOfTariffsUnsorted
    * @return 2 lists: the sorted List<String> tariffsToDisplay and the sorted List<TariffVO> tariffVOs
    */
``` 

which causes a parse error because of unclosed HTML tag, which of course means that in this case the angle brackets should be escaped correctly.

Later in the analysis there is a file B.java with the following JavaDoc:

```
   /**
    * @see javax.swing.tree.DefaultTreeCellRenderer.getTreeCellRendererComponent()
    */
```

which causes a syntax error because of the wrong @see reference which should of course be `@see javax.swing.tree.DefaultTreeCellRenderer#getTreeCellRendererComponent` instead.

 I don't have an error or stacktrace, besides the subsequent error in SonarQube, but that is not really Checkstyle related. The effect in Checkstyle is, that you get the wrong finding for the second file. 

**Example:**
*Detected:*
```
java -cp 'checkstyle-6.11.2-all.jar' com.puppycrawl.tools.checkstyle.Main -c checkstyle.xml A.java B.java
Starting audit...
...\A.java:1024: warning: Javadoc comment at column 84 has parse error. Missed HTML close tag 'TariffVO'. Sometimes it means that close tag missed for one of previous tags.
...\B.java:1024: warning: Javadoc comment at column 84 has parse error. Missed HTML close tag 'TariffVO'. Sometimes it means that close tag missed for one of previous tags.
Audit done.
```
*Expected:*
```
java -cp 'target\classes;checkstyle-6.11.2-all.jar' com.puppycrawl.tools.checkstyle.Main -c checkstyle.xml A.java B.java
Starting audit...
...\A.java:1024: warning: Javadoc comment at column 84 has parse error. Missed HTML close tag 'TariffVO'. Sometimes it means that close tag missed for one of previous tags.
...\B.java:26: warning: Javadoc comment at column 3 has parse error. Unrecognized error from ANTLR parser: null
Audit done.
```

As it seems SQ is reusing the same instance of AtclauseOrderCheck for multiple files to be analyzed.
Because of this the following happens:

The error in file B.java is recognized as syntax error, but not handled by [AbstractJavadocCheck.DescriptiveErrorListener#syntaxError](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/AbstractJavadocCheck.java#L605), but by BailErrorStrategy#recoverInline([line](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/AbstractJavadocCheck.java#L506)). This should be handled by AbstractJavadocCheck lines [236 - 241](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/AbstractJavadocCheck.java#L236-L242) in current master file I think, but the DescriptiveErrorListener still has the last error message found which is used, instead of the new one. In this case it is the error message from A.java.

This causes a wrong message and line information, as line [237](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/AbstractJavadocCheck.java#L237) is false.
In our case the error in A.java is in line 1024 and the file B.java has only 42 lines.
this later on causes and ArrayIndexOutOfBoundsException in the SonarQube analysis.

**Comment by Vampire** [comment]
I created a pull request which fixes this issue.

**Comment by romani** [comment]
@Vampire , thanks for reporting an issue.

1) Please share a error or stacktrace that you have

2) Please read your description again and update it to have more details, looks like you miss paragraph before "As it seems SQ is reusing the same". Please focus on details of what problem you have rather then on how to fix it.

3) Please reproduce a problem by our CLI (http://checkstyle.sourceforge.net/cmdline.html#Download_and_Run), put only two files to validate.

4) for PR: Checkstyle is using Test Driven Development, so we need to reproduce it by UT.

**Comment by Vampire** [comment]
1) I don't have an error or stacktrace, besides the subsequent error in SonarQube, but that is not really Checkstyle related. The effect in Checkstyle is, that you get the wrong finding for the second file. 

**Example:**
*Wrong:*
```
java -cp 'checkstyle-6.11.2-all.jar' com.puppycrawl.tools.checkstyle.Main -c checkstyle.xml A.java B.java
Starting audit...
...\A.java:1024: warning: Javadoc comment at column 84 has parse error. Missed HTML close tag 'TariffVO'. Sometimes it means that close tag missed for one of previous tags.
...\B.java:1024: warning: Javadoc comment at column 84 has parse error. Missed HTML close tag 'TariffVO'. Sometimes it means that close tag missed for one of previous tags.
Audit done.
```
*Correct:*
```
java -cp 'target\classes;checkstyle-6.11.2-all.jar' com.puppycrawl.tools.checkstyle.Main -c checkstyle.xml A.java B.java
Starting audit...
...\A.java:1024: warning: Javadoc comment at column 84 has parse error. Missed HTML close tag 'TariffVO'. Sometimes it means that close tag missed for one of previous tags.
...\B.java:26: warning: Javadoc comment at column 3 has parse error. Unrecognized error from ANTLR parser: null
Audit done.
```

2) Well as far as I can see I clearly described the situation we have, the problem we have and proposed a possible cause. I don't really see what is unclear. Please ask specific questions if you don't understand something.

3) I did before I reported and at 1) you see the result

4) You are welcome to do it. I tried to write a test first, but your BaseCheckTestSupport does not support the use-case. You can give it multiple files to validate, but you cannot only check for multiple findings in one file and not in mutliple files. As I fixed the bug in my little spare-time and didn't want to rewrite your test framework to support the test case, I just submitted the manually tested fix for the problem. If you have other facilities to create the test, be my guest.

**Comment by romani** [comment]
@baratali , can you comment on this issue and PR ?

**Comment by Vampire** [comment]
Ok, I took some minutes and gave it another try.
Actually in master it was easier than in the 6.4.1 tag, or a fresh look made it easier.
Whatever, I added test coverage now.

**Comment by romani** [comment]
@Vampire , Sorry for late focus on your issue/pr, we support whole project in our "little spare-time", I put a priority to issues that are completely clear and ready with PRs. We have a lot of proposed PRs per day.

I need to review our old intentions (there were some nuances) and your fix, it will takes some time. On first site your change is reasonable, but let me recheck.

**Comment by Vampire** [comment]
Sure, no problem. :-)

**Comment by baratali** [comment]
Looks like it can be reproduced on a single file. For example:
```java
public class A {
   /**
    * get the sorted list of tariffs to display in the combo box
    *
    * @param listOfTariffsUnsorted
    * @return 2 lists: the sorted List<String> tariffsToDisplay and the sorted List<TariffVO> tariffVOs
    */
    public void a() {}

  /**
    * @see javax.swing.tree.DefaultTreeCellRenderer.getTreeCellRendererComponent()
    */
    public void b() {}
}
```

I'll test it today...

**Comment by Vampire** [comment]
Yeah, actually this should reproduce the probelm also, you are right. Do you prefer that the test works with just that one file and my changes about being able to test with errors from multiple files reverted?

**Comment by baratali** [comment]
It is reproduced on single file. However, only one message is printed. I guess, the same errors is logged by checkstyle.
```
bizmailov@vdu-bizmailov:~/git/checkstyle/target$ java -cp 'checkstyle-6.12-SNAPSHOT-all.jar' com.puppycrawl.tools.checkstyle.Main -c checkstyle_checks.xml A.java
Starting audit...
/home/bizmailov/git/checkstyle/target/A.java:6: error: Javadoc comment at column 84 has parse error. Missed HTML close tag 'TariffVO'. Sometimes it means that close tag missed for one of previous tags.
Audit done.
Checkstyle ends with 1 errors.
```

@Vampire , I don't mind about two-files test. This PR should resolve this issue with single file as well.

**Comment by Vampire** [comment]
I think so, yes

**Comment by Vampire** [comment]
Yes, it fixed it also. I also added the one-file check to my PR as I had to rework and rebase it anyway.

**Comment by romani** [comment]
@Vampire , thank you very mush for your time.

> Do you prefer that the test works with just that one file and my changes about being able to test with errors from multiple files reverted?

lets keep both tests, the more tests the better, and your tests are a bit different.

last required updates:
1) 
> OrderedMap --> Map

Keeping warning in Javadoc never works well - code should force engineer to do things in right way.
And we are almost done with removing dependency to commons-lang3 .
Lets look on a problem from different side - we should not demand from author of UTs order of violations and order of files. It is not a contract of Checkstyle. Lets update a bit ```protected void verify(Checker checker, File[] processedFiles, Map<String, String[]> expectedMessages)``` and store actual messages to Map<name_of_fIle,array_of_violations> and do intersection. For now we could rely on order of violations in one file.
Does it make sense ? If "yes" please do.

2)
> ```Entry<String, String[]> expectedMessage```

please rename it to expectedViolations

3)
> final String[] expected

final String[] expectedMessages

4)
> final String actual

final String actualResult

**Comment by romani** [comment]
discussion is at #2328
